File: js-packages/misc/src/utils/base_crypto.ts
import {
    unserializeToArrayBuffer,
    serializeToBase64,
    splitFirstOnly,
    splitFirstOnlyInverted,
} from './encoding'
import { ValueType } from '../typing'
import { unserializeToCryptoKey } from './base_crypto_legacy'

export function addWithVariants<T extends { [key: string]: any }>(
    obj: T,
    newEntry: ValueType<T>,
    variants: string[]
) {
    for (const variant of variants) {
        ;(obj as any)[variant] = newEntry
    }
}

export let DEFAULT_SIGNATURE_ALGORITHM = 'rsa-sha512'
export let DEFAULT_ASYMMETRIC_ENCRYPTION_ALGORITHM = 'rsa-sha512'
export let DEFAULT_SYMMETRIC_ENCRYPTION_ALGORITHM = 'AESGCM'
export let DEFAULT_DERIVE_ALGORITHM = 'PBKDF2-sha512'

export type ParamsType = any
export type KeyType = CryptoKey | ArrayBuffer

export interface DeriveResult {
    data: ArrayBuffer
    params: ParamsType
}
export interface CryptoResult extends DeriveResult {
    key: ArrayBuffer
}

export interface KeyResult {
    key: ArrayBuffer
    params: ParamsType
}

export interface DeserializeResult {
    data?: ArrayBuffer
    params: ParamsType
}

export interface DeserializeDeriveResult extends DeserializeResult {
    data: ArrayBuffer
}

export abstract class DeriveAlgorithm {
    serializedName: string
    type: 'hash' | 'derive'
    abstract derive(
        inp: ArrayBuffer,
        params?: ParamsType
    ): Promise<DeriveResult>
    async deserialize(
        inp: string,
        params?: ParamsType
    ): Promise<DeserializeDeriveResult> {
        const match = inp.match(/^:*(.*?)$/) as string[]
        return {
            data: await unserializeToArrayBuffer(match[1]),
            params: params || {},
        }
    }
    async serialize(inp: {
        data: ArrayBuffer
        params: ParamsType
    }): Promise<string> {
        return await serializeToBase64(inp.data)
    }
}

export abstract class EncryptionAlgorithm {
    serializedName: string
    type: 'asymmetric' | 'symmetric'

    abstract encrypt(
        key: KeyType,
        data: ArrayBuffer,
        params: ParamsType
    ): Promise<CryptoResult>
    abstract decrypt(
        key: KeyType,
        data: ArrayBuffer,
        params: ParamsType
    ): Promise<CryptoResult>
    abstract generateKey(params: ParamsType): Promise<KeyResult>
    abstract toHashableKey(key: CryptoKey | ArrayBuffer): Promise<ArrayBuffer>
    async serializeParams(params: ParamsType): Promise<string> {
        return ''
    }
    async deserialize(
        inp: string,
        params?: ParamsType
    ): Promise<DeserializeResult> {
        const match = inp.match(/^:*(.*?)$/) as string[]
        return {
            data: await unserializeToArrayBuffer(match[1]),
            params: params || {},
        }
    }
}

export abstract class SignatureAlgorithm {
    serializedName: string
    abstract sign(key: KeyType, data: ArrayBuffer): Promise<string>
    abstract verify(
        key: KeyType,
        signature: string,
        data: ArrayBuffer
    ): Promise<boolean>
    abstract generateKey(params: ParamsType): Promise<KeyResult>
    abstract toHashableKey(key: CryptoKey | ArrayBuffer): Promise<ArrayBuffer>
}

// argon1id, pkb
export const mapDeriveAlgorithms: {
    [algo: string]: DeriveAlgorithm
} = {}

export const mapEncryptionAlgorithms: {
    [algo: string]: EncryptionAlgorithm
} = {}
export const mapSignatureAlgorithms: {
    [algo: string]: SignatureAlgorithm
} = {}

class HashAlgos extends DeriveAlgorithm {
    type = 'hash' as const
    operationName: string
    constructor(operationName: string) {
        super()
        this.operationName = operationName
        this.serializedName = operationName.toLowerCase().replace('-', '')
    }
    async derive(inp: ArrayBuffer) {
        return {
            data: await crypto.subtle.digest(this.operationName, inp),
            params: {},
        }
    }
}
addWithVariants(mapDeriveAlgorithms, new HashAlgos('SHA-256'), [
    'sha256',
    'SHA-256',
])

addWithVariants(mapDeriveAlgorithms, new HashAlgos('SHA-512'), [
    'sha512',
    'SHA-512',
])

class PBKDF2Algos extends DeriveAlgorithm {
    type = 'derive' as const
    operationName: string
    constructor(operationName: string) {
        super()
        this.operationName = operationName
        this.serializedName = `PBKDF2-${operationName
            .toLowerCase()
            .replace('-', '')}`
    }
    async derive(
        inp: ArrayBuffer,
        params: {
            iterations: number
            salt?: string
        } = {
            iterations: 800000,
        }
    ): Promise<CryptoResult> {
        // copy
        params = Object.assign({}, params)
        const key = await crypto.subtle.importKey(
            'raw',
            inp,
            'PBKDF2',
            false,
            ['deriveBits', 'deriveKey']
        )
        if (!params.salt) {
            throw Error('no salt provided')
        }
        const salt = await unserializeToArrayBuffer(params.salt as any)
        return {
            data: await crypto.subtle.deriveBits(
                {
                    name: 'PBKDF2',
                    salt,
                    iterations: params.iterations,
                    hash: 'SHA-512',
                },
                key,
                256 // cap at 256 for AESGCM compatibility
            ),
            params,
            key: inp,
        }
    }
    async deserialize(
        inp: string,
        params: {
            iterations: number
            salt?: string
        } = {
            iterations: 800000,
        }
    ) {
        params = Object.assign({}, params)
        const splitted = inp.split(':')
        if (splitted.length >= 2) {
            const splitted2 = splitted[0].split(',')
            params.iterations = parseInt(splitted2[0])
            if (splitted2.length > 2) {
                params.salt = splitted2[1]
            }
        }
        if (!params.salt) {
            throw Error('no salt provided')
        }
        const data = await unserializeToArrayBuffer(
            splitted[splitted.length - 1]
        )
        return { data, params }
    }
    async serialize(inp: CryptoResult) {
        return `${inp.params.iterations},${
            inp.params.salt
        }:${await serializeToBase64(inp.data)}`
    }
}
addWithVariants(mapDeriveAlgorithms, new PBKDF2Algos('SHA-512'), [
    'PBKDF2-sha512',
])
addWithVariants(mapDeriveAlgorithms, new PBKDF2Algos('SHA-256'), [
    'PBKDF2-256',
])

class RSAOEAPAlgos extends EncryptionAlgorithm {
    type = 'asymmetric' as const
    operationName: string
    declare keyParams: {
        name: 'RSA-OAEP'
        hash: string
    }
    constructor(operationName: string) {
        super()
        this.keyParams = {
            name: 'RSA-OAEP',
            hash: operationName,
        }
        this.serializedName = `rsa-${operationName
            .toLowerCase()
            .replace('-', '')}`
    }
    async encrypt(key: KeyType, data: ArrayBuffer) {
        const key_cleaned = await unserializeToCryptoKey(
            key,
            this.keyParams,
            'publicKey'
        )
        return {
            data: await crypto.subtle.encrypt(
                {
                    name: 'RSA-OAEP',
                },
                key_cleaned,
                data
            ),
            params: {},
            key: await unserializeToArrayBuffer(key_cleaned),
        }
    }
    async decrypt(key: KeyType, data: ArrayBuffer) {
        const key_cleaned = await unserializeToCryptoKey(
            key,
            this.keyParams,
            'privateKey'
        )
        return {
            data: await crypto.subtle.decrypt(
                {
                    name: 'RSA-OAEP',
                },
                key_cleaned,
                data
            ),
            params: {},
            key: await unserializeToArrayBuffer(key_cleaned),
        }
    }
    async toHashableKey(key: CryptoKey | ArrayBuffer): Promise<ArrayBuffer> {
        const publicKey = await unserializeToCryptoKey(
            key,
            this.keyParams,
            'publicKey'
        )
        return await crypto.subtle.exportKey('spki' as const, publicKey)
    }
    async generateKey({ bits }: { bits: number } = { bits: 4096 }) {
        const keypair = await crypto.subtle.generateKey(
            {
                modulusLength: bits,
                publicExponent: new Uint8Array([1, 0, 1]),
                ...this.keyParams,
            },
            true,
            ['wrapKey', 'unwrapKey', 'encrypt', 'decrypt']
        )
        return {
            key: await unserializeToArrayBuffer(keypair.privateKey),
            params: { bits },
        }
    }
}
addWithVariants(mapEncryptionAlgorithms, new RSAOEAPAlgos('SHA-512'), [
    'rsa-sha512',
    'sha512',
    'SHA-512',
])

addWithVariants(mapEncryptionAlgorithms, new RSAOEAPAlgos('SHA-256'), [
    'rsa-sha256',
    'sha256',
    'SHA-256',
])

class AESGCMAlgos extends EncryptionAlgorithm {
    serializedName = 'AESGCM'
    type = 'symmetric' as const
    keyParams = {
        name: 'AES-GCM',
    }
    async encrypt(
        key: KeyType,
        data: ArrayBuffer,
        params?: { nonce: string | ArrayBuffer }
    ) {
        const key_cleaned = await unserializeToCryptoKey(key, this.keyParams)
        let nonce
        if (params?.nonce) {
            nonce = await unserializeToArrayBuffer(params.nonce)
        } else {
            nonce = crypto.getRandomValues(new Uint8Array(13))
        }
        return {
            data: await crypto.subtle.encrypt(
                {
                    name: 'AES-GCM',
                    iv: nonce,
                },
                key_cleaned,
                data
            ),
            params: { nonce },
            key: await unserializeToArrayBuffer(key_cleaned),
        }
    }
    async decrypt(
        key: KeyType,
        data: ArrayBuffer,
        params: { nonce: string | ArrayBuffer }
    ) {
        if (!params?.nonce) {
            throw new Error('missing nonce')
        }
        const key_cleaned = await unserializeToCryptoKey(key, this.keyParams)
        const nonce = await unserializeToArrayBuffer(params.nonce)
        return {
            data: await crypto.subtle.decrypt(
                {
                    name: 'AES-GCM',
                    iv: nonce,
                },
                key_cleaned,
                data
            ),
            params: { nonce },
            key: await unserializeToArrayBuffer(key_cleaned),
        }
    }
    async serializeParams(params: { nonce: ArrayBuffer }) {
        return `${await serializeToBase64(params.nonce)}`
    }
    async deserialize(inp: string, params?: { nonce: ArrayBuffer }) {
        let params2: { nonce?: ArrayBuffer } = Object.assign({}, params || {})
        let splitted = splitFirstOnlyInverted(inp)
        params2.nonce = await unserializeToArrayBuffer(splitted[0])
        let cleaned: ArrayBuffer | undefined = undefined
        if (params2.nonce && splitted[splitted.length - 1]) {
            cleaned = await unserializeToArrayBuffer(
                splitted[splitted.length - 1]
            )
        }
        return {
            params: params2,
            data: cleaned,
        }
    }
    async toHashableKey(key: ArrayBuffer): Promise<ArrayBuffer> {
        if (!(key instanceof ArrayBuffer)) {
            throw Error('invalid key type')
        }
        // already strengthed
        if (key.byteLength >= 50) {
            return key
        }
        if (key.byteLength != 32) {
            throw Error('invalid key length for hashing')
        }
        const prefix = crypto.getRandomValues(new Uint8Array(18))
        const fullArray = new Uint8Array(50)
        fullArray.set(prefix)
        fullArray.set(new Uint8Array(key), 18)
        return fullArray
    }
    async generateKey() {
        return {
            key: crypto.getRandomValues(new Uint8Array(32)),
            params: {},
        }
    }
}
addWithVariants(mapEncryptionAlgorithms, new AESGCMAlgos(), ['AESGCM'])

class RSAPSSAlgos extends SignatureAlgorithm {
    type = 'asymmetric' as const
    operationName: string
    saltLength: number
    declare keyParams: {
        name: 'RSA-PSS'
        hash: string
    }
    constructor(operationName: string, saltLength: number) {
        super()
        this.saltLength = saltLength
        this.keyParams = {
            name: 'RSA-PSS',
            hash: operationName,
        }
        this.serializedName = `rsa-${operationName
            .toLowerCase()
            .replace('-', '')}`
    }
    async sign(key: KeyType, data: ArrayBuffer) {
        const key_cleaned = await unserializeToCryptoKey(
            key,
            this.keyParams,
            'privateKey'
        )
        return await serializeToBase64(
            crypto.subtle.sign(
                {
                    name: 'RSA-PSS',
                    saltLength: this.saltLength,
                },
                key_cleaned,
                data
            )
        )
    }

    async verify(key: KeyType, signature: string, data: ArrayBuffer) {
        const key_cleaned = await unserializeToCryptoKey(
            key,
            this.keyParams,
            'publicKey'
        )
        return await crypto.subtle.verify(
            {
                name: 'RSA-PSS',
                saltLength: this.saltLength,
            },
            key_cleaned,
            await unserializeToArrayBuffer(signature),
            data
        )
    }
    async generateKey({ bits }: { bits: number } = { bits: 4096 }) {
        const keypair = await crypto.subtle.generateKey(
            {
                modulusLength: bits,
                publicExponent: new Uint8Array([1, 0, 1]),
                ...this.keyParams,
            },
            true,
            ['wrapKey', 'unwrapKey', 'encrypt', 'decrypt']
        )
        return {
            key: await unserializeToArrayBuffer(keypair.privateKey),
            params: { bits },
        }
    }
    async toHashableKey(key: CryptoKey | ArrayBuffer): Promise<ArrayBuffer> {
        const publicKey = await unserializeToCryptoKey(
            key,
            this.keyParams,
            'publicKey'
        )
        return await crypto.subtle.exportKey('spki' as const, publicKey)
    }
}
addWithVariants(mapSignatureAlgorithms, new RSAPSSAlgos('SHA-512', 64), [
    'rsa-sha512',
    'sha512',
    'SHA-512',
])
addWithVariants(mapSignatureAlgorithms, new RSAPSSAlgos('SHA-256', 32), [
    'rsa-sha256',
    'sha256',
    'SHA-256',
])

File: secretgraph/core/utils/base_crypto.py
import asyncio
import copy
import hashlib
from abc import ABC, abstractmethod
from base64 import b64decode, b64encode
from collections.abc import Callable
from dataclasses import dataclass, field
from inspect import isclass
from os import urandom
from typing import Any, Iterable, Literal, NoReturn, Optional, TypeVar

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa, utils
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.hashes import Hash

T = TypeVar("T")
KeyType = TypeVar("KeyType", bound=bytes)
ParamsType = TypeVar("ParamsType", bound=dict)


def addWithVariants(
    targetDict: dict[str, T], variants: list[str], entry: Optional[T] = None
):
    def wrapper(entry):
        if isclass(entry):
            initialized = entry()
        else:
            initialized = entry
        for variant in variants:
            targetDict[variant] = initialized
        return entry

    if entry is not None:
        return wrapper(entry)
    return wrapper


@dataclass(frozen=True, kw_only=True)
class DeriveResult:
    data: bytes
    params: ParamsType = field(default_factory=dict)


@dataclass(frozen=True, kw_only=True)
class CryptoResult(DeriveResult):
    key: bytes


@dataclass(frozen=True, kw_only=True)
class KeyResult:
    key: KeyType
    params: ParamsType = field(default_factory=dict)


@dataclass(frozen=True, kw_only=True)
class DeserializeResult:
    params: ParamsType
    data: Optional[bytes] = None


@dataclass(frozen=True, kw_only=True)
class DeserializeDeriveResult(DeserializeResult):
    data: bytes


@dataclass(frozen=True, kw_only=True)
class Hasher:
    update: Callable[[bytes], NoReturn]
    finalize: Callable[[], bytes]


class DeriveAlgorithm(ABC):
    serializedName: str
    type: Literal["hash"] | Literal["derive"]

    @abstractmethod
    async def derive(
        self, data: bytes | Iterable[bytes], params: Optional[ParamsType] = None
    ) -> CryptoResult:
        pass

    async def serialize(self, result: DeriveResult) -> str:
        return b64encode(result.data).decode()

    async def deserialize(
        self, inp: str, params: Optional[ParamsType] = None
    ) -> DeserializeDeriveResult:
        return DeserializeDeriveResult(data=b64decode(inp), params=params or {})


class EncryptionAlgorithm(ABC):
    serializedName: str
    type: Literal["asymmetric"] | Literal["symmetric"]

    @abstractmethod
    async def encrypt(
        self, key: KeyType, data: bytes, params: ParamsType
    ) -> CryptoResult:
        pass

    async def encryptor(self, key: KeyType, params: ParamsType):
        raise NotImplementedError()

    @abstractmethod
    async def decrypt(
        self, key: KeyType, data: bytes, params: ParamsType
    ) -> CryptoResult:
        pass

    async def decryptor(self, key: KeyType, params: ParamsType) -> CryptoResult:
        raise NotImplementedError()

    @abstractmethod
    def toHashableKey(self, key: KeyType, raw: bool) -> bytes | Any:
        pass

    @abstractmethod
    async def generateKey(self, params: ParamsType) -> KeyType:
        pass

    async def serializeParams(self, result: CryptoResult) -> str:
        return ""

    async def deserialize(self, data: str, params: ParamsType) -> DeserializeResult:
        return DeserializeResult(data=b64decode(data), params=params or {})


class SignatureAlgorithm(ABC):
    serializedName: str

    @abstractmethod
    async def sign(self, key: KeyType, inp: bytes, prehashed: bool = False) -> str:
        pass

    @abstractmethod
    async def verify(
        self, key: KeyType, signature: str, inp: bytes, prehashed: bool = False
    ) -> bool:
        pass

    async def getHasher(self) -> Hasher:
        raise NotImplementedError()

    @abstractmethod
    def toHashableKey(self, key: KeyType, raw: bool) -> bytes | Any:
        pass

    @abstractmethod
    async def generateKey(self, params: ParamsType) -> KeyType:
        pass


mapDeriveAlgorithms: dict[str, DeriveAlgorithm] = {}
mapEncryptionAlgorithms: dict[str, EncryptionAlgorithm] = {}
mapSignatureAlgorithms: dict[str, SignatureAlgorithm] = {}


@addWithVariants(mapDeriveAlgorithms, ["sha512"])
class SHA512Algo(DeriveAlgorithm):
    type = "hash"
    serializedName = "sha512"

    def _execute(self, data: bytes | Iterable[bytes]):
        hashCtx = hashlib.new(self.serializedName)
        if isinstance(data, bytes):
            hashCtx.update(data)
        else:
            for chunk in data:
                hashCtx.update(chunk)
        return hashCtx.digest()

    async def execute(self, data: bytes | Iterable[bytes]):
        return await asyncio.get_event_loop().run_in_executor(None, self._execute, data)

    async def derive(self, inp: bytes | Iterable[bytes], params=None) -> CryptoResult:
        return DeriveResult(data=await self.execute(inp), params={})

    async def serialize(self, result: CryptoResult) -> str:
        return b64encode(result.data).decode()


@addWithVariants(mapDeriveAlgorithms, ["sha256"])
class SHA256Algo(SHA512Algo):
    serializedName = "sha256"


@addWithVariants(mapDeriveAlgorithms, ["PBKDF2-sha512"])
class PBKDF2sha512(DeriveAlgorithm):
    type = "derive"
    serializedName = "PBKDF2-sha512"

    def _execute(self, iterations, salt, data):
        if not isinstance(data, bytes):
            data = b"".join(data)
        return hashlib.pbkdf2_hmac(
            self.serializedName.split("-")[1],
            data,
            iterations=iterations,
            salt=salt,
            dklen=32,
        ).digest()

    async def execute(self, iterations, salt, data):
        return await asyncio.get_event_loop().run_in_executor(
            None, self._execute, iterations, salt, data
        )

    async def derive(self, inp: bytes, params=None) -> CryptoResult:
        if not params:
            params = {}
        else:
            params = copy.copy(params)
        params["iterations"] = int(params.get("iterations", 800000))
        params["salt"] = params.get("salt", None)
        if not params["salt"]:
            params["salt"] = urandom(16)
        if isinstance(params["salt"], str):
            params["salt"] = b64decode(params["salt"])

        # for AESGCM compatibility cap at 32
        return DeriveResult(
            data=await self.execute(params["iterations"], params["salt"], inp),
            params=params,
        )

    async def deserialize(self, inp: str, params=None):
        if not params:
            params = {}
        else:
            params = copy.copy(params)
        splitted = inp.split(":")
        if splitted.length >= 2:
            splitted2 = splitted[0].split(",")
            params["iterations"] = int(splitted2[0])
            if splitted2.length > 2:
                params["salt"] = splitted2[1]

        if not params.get("salt"):
            raise ValueError("no salt provided")
        if isinstance(params["salt"], str):
            params["salt"] = b64decode(params["salt"])
        return DeserializeDeriveResult(data=b64decode(splitted[-1]), params=params)

    async def serialize(self, result):
        return f"""{result.params['iterations']},{
            result.params['salt']
        }:${b64encode(result.data).decode()}"""


@addWithVariants(mapDeriveAlgorithms, ["PBKDF2-sha512"])
class PBKDF2sha256(PBKDF2sha512):
    serializedName = "PBKDF2-sha256"


@addWithVariants(mapEncryptionAlgorithms, ["rsa-sha512"])
class OEAPsha512(EncryptionAlgorithm):
    type = "asymmetric"
    serializedName = "rsa-sha512"

    async def encrypt(self, key, data, params=None):
        # to publicKey
        key = await self.toHashableKey(key, raw=True)
        hashalgo = self.serializedName.split("-")[1].upper()
        hashalgo = getattr(hashes, hashalgo)()
        return CryptoResult(
            data=key.encrypt(
                data,
                padding.OAEP(
                    mgf=padding.MGF1(algorithm=hashalgo),
                    algorithm=hashalgo,
                    label=None,
                ),
            ),
            key=key.public_bytes(
                encoding=serialization.Encoding.DER,
                format=serialization.PublicFormat.SubjectPublicKeyInfo,
            ),
        )

    async def toHashableKey(self, key, raw):
        if isinstance(key, str):
            key = b64decode(key)
        if isinstance(key, bytes):
            try:
                key = serialization.load_der_public_key(key)
            except Exception:
                key = serialization.load_der_private_key(key, None)
        if hasattr(key, "public_key"):
            key = key.public_key()
        if raw:
            return key
        return key.public_bytes(
            encoding=serialization.Encoding.DER,
            format=serialization.PublicFormat.SubjectPublicKeyInfo,
        )

    async def decrypt(self, key, data, params=None):
        if isinstance(key, str):
            key = b64decode(key)
        if isinstance(key, bytes):
            key = serialization.load_der_private_key(key, None)
        hashalgo = self.serializedName.split("-")[1].upper()
        hashalgo = getattr(hashes, hashalgo)()
        return CryptoResult(
            data=key.decrypt(
                data,
                padding.OAEP(
                    mgf=padding.MGF1(algorithm=hashalgo),
                    algorithm=hashalgo,
                    label=None,
                ),
            ),
            key=key.private_bytes(
                encoding=serialization.Encoding.DER,
                format=serialization.PrivateFormat.PKCS8,
                encryption_algorithm=serialization.NoEncryption(),
            ),
        )

    async def generateKey(cls, params=None):
        if not params:
            params = {}
        else:
            params = copy.copy(params)
        params["bits"] = int(params.get("bits", 4096))
        return KeyResult(
            key=rsa.generate_private_key(
                public_exponent=65537,
                key_size=params["bits"],
            ).private_bytes(
                encoding=serialization.Encoding.DER,
                format=serialization.PrivateFormat.PKCS8,
                encryption_algorithm=serialization.NoEncryption(),
            ),
            params=params,
        )


@addWithVariants(mapEncryptionAlgorithms, ["rsa-sha256"])
class OEAPsha256(OEAPsha512):
    serializedName = "rsa-sha256"


@addWithVariants(mapEncryptionAlgorithms, ["AESGCM"])
class AESGCMAlgo(EncryptionAlgorithm):
    type = "symmetric"
    serializedName = "AESGCM"

    async def encrypt(self, key, data, params=None):
        if not params:
            params = {}
        else:
            params = copy.copy(params)
        if isinstance(key, str):
            key = b64decode(key)
        if isinstance(key, bytes):
            key = AESGCM(key)
        if not params.get("nonce"):
            params["nonce"] = urandom(13)
        if isinstance(params["nonce"], str):
            params["nonce"] = b64decode(params["nonce"])
        return CryptoResult(
            data=key.encrypt(params["nonce"], data, None), params=params, key=key._key
        )

    async def encryptor(self, key, params):
        params = copy.copy(params)
        if isinstance(params["nonce"], str):
            params["nonce"] = b64decode(params["nonce"])
        if isinstance(key, str):
            key = b64decode(key)
        return Cipher(
            algorithms.AES(key),
            modes.GCM(params["nonce"]),
        ).encryptor()

    async def decrypt(self, key, data, params):
        params = copy.copy(params)
        if isinstance(params["nonce"], str):
            params["nonce"] = b64decode(params["nonce"])
        if isinstance(key, str):
            key = b64decode(key)
        if isinstance(key, bytes):
            key = AESGCM(key)
        return CryptoResult(
            data=key.decrypt(params["nonce"], data, None), params=params, key=key._key
        )

    async def decryptor(self, key, params):
        params = copy.copy(params)
        if isinstance(params["nonce"], str):
            params["nonce"] = b64decode(params["nonce"])
        if isinstance(key, str):
            key = b64decode(key)
        return Cipher(
            algorithms.AES(key),
            modes.GCM(params["nonce"]),
        ).decryptor()

    async def serializeParams(self, params):
        return b64encode(params["nonce"]).decode()

    async def toHashableKey(self, key: KeyType, raw):
        # already strengthed
        if len(key) >= 50:
            return key
        if len(key) != 32:
            raise ValueError("invalid key length for hashing")
        return urandom(18) + key

    async def generateKey(self, params=None):
        if not params:
            params = {}
        else:
            params = copy.copy(params)
        params["bits"] = int(params.get("bits", 256))
        return KeyResult(
            key=AESGCM.generate_key(params["bits"]),
            params=params,
        )


@addWithVariants(mapSignatureAlgorithms, ["rsa-sha512", "sha512", "SHA-512"])
class RSASignsha512(SignatureAlgorithm):
    serializedName = "rsa-sha512"
    generateKey = OEAPsha512.generateKey
    toHashableKey = OEAPsha512.toHashableKey

    async def sign(self, key, data, prehashed=False):
        if isinstance(key, str):
            key = b64decode(key)
        if isinstance(key, bytes):
            key = serialization.load_der_private_key(key, None)
        hashalgo = self.serializedName.split("-")[1].upper()
        hashalgo = getattr(hashes, hashalgo)()
        if prehashed:
            hash2 = utils.Prehashed(hashalgo)
        else:
            hash2 = hashalgo
        return b64encode(
            key.sign(
                data,
                padding.PSS(
                    mgf=padding.MGF1(hashalgo), salt_length=padding.PSS.MAX_LENGTH
                ),
                hash2,
            )
        ).decode()

    async def getHasher(self):
        hashalgo = self.serializedName.split("-")[1].upper()
        hashalgo = getattr(hashes, hashalgo)()
        hashCtx = Hash(hashalgo)
        return Hasher(update=hashCtx.update, finalize=hashCtx.finalize)

    async def verify(self, key, signature, data, prehashed=False):
        key = await self.toHashableKey(key, raw=True)
        if isinstance(signature, str):
            signature = b64decode(signature)
        if isinstance(data, str):
            data = b64decode(data)
        hashalgo = self.serializedName.split("-")[1].upper()
        hashalgo = getattr(hashes, hashalgo)()
        if prehashed:
            hash2 = utils.Prehashed(hashalgo)
        else:
            hash2 = hashalgo
        try:
            key.verify(
                signature,
                data,
                padding.PSS(
                    mgf=padding.MGF1(hashalgo), salt_length=padding.PSS.MAX_LENGTH
                ),
                hash2,
            )
            return True
        except InvalidSignature:
            return False


@addWithVariants(mapSignatureAlgorithms, ["rsa-sha256", "sha256", "SHA-256"])
class RSASignsha256(RSASignsha512):
    serializedName = "rsa-sha256"
